src/deepMerge.ts

```diff
@@ -0,0 +1,112 @@
+import type { MergeDeep } from 'type-fest';
+import { purry } from './purry';
+
+type UnknownRecordOrArray = Record<string, unknown> | Array<unknown>;
+
+function isRecordOrArray(object: unknown): object is UnknownRecordOrArray {
+  return typeof object === 'object' && object !== null;
+}
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ * @param target - The first value to merge. This is the dominant parameter in the merge operation.
+ * @param source - The second value to merge.
+ * @signature
+ *    R.deepMerge(a, b)
+ * @example
+ *    R.deepMerge({ foo: 'bar', x: 1 }, { foo: 'baz', y: 2 }) // => { foo: 'bar', x: 1, y: 2 }
+ * @data_first
+ * @category Object
+ * @pipeable
+ */
+export function deepMerge<
+  Target extends UnknownRecordOrArray,
+  Source extends UnknownRecordOrArray
+>(target: Target, source: Source): MergeDeep<Target, Source>;
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ * @param source - The second value to merge.
+ * @signature
+ *    R.deepMerge(b)(a)
+ * @example
+ *    R.deepMerge({ foo: 'baz', y: 2 })({ foo: 'bar', x: 1 }) // => { foo: 'bar', x: 1, y: 2 }
+ *    R.pipe(
+ *      { foo: 'bar', x: 1 },
+ *      R.deepMerge({ foo: 'baz', y: 2 }),
+ *    ) // => { foo: 'bar', x: 1, y: 2 }
+ * @data_last
+ * @category Object
+ * @pipeable
+ */
+export function deepMerge<
+  Target extends UnknownRecordOrArray,
+  Source extends UnknownRecordOrArray
+>(source: Source): (target: Target) => MergeDeep<Target, Source>;
+
+export function deepMerge() {
+  return purry(_deepMerge, arguments);
+}
+
+export function _deepMerge(a: UnknownRecordOrArray, b: UnknownRecordOrArray) {
+  const seenValues = new WeakSet(); // use to keep track of which objects have been seen.
+
+  function isArrayCyclic(arr: Array<unknown>) {
+    return arr.some(value => value && seenValues.has(value));
+  }
+
+  function merge(a: UnknownRecordOrArray, b: UnknownRecordOrArray) {
+    if (
+      typeof a !== 'object' ||
+      typeof b !== 'object' ||
+      a === null ||
+      b === null
+    ) {
+      return a ?? b;
+    }
+
+    if (seenValues.has(a) || seenValues.has(b)) {
+      return a;
+    }
+    seenValues.add(a);
+    seenValues.add(b);
+
+    if (Array.isArray(a)) {
+      if (Array.isArray(b)) {
+        if (isArrayCyclic(a)) {
+          return b;
+        }
+        if (isArrayCyclic(b)) {
+          return a;
+        }
```

Comment:
~~Can you add documentation here, I'm not sure I understand why if one array contains a recursive element then the merge result is simply the other array...~~

EDIT: Thinking about this again, because we don't deep merge the array items I think it's best to simply not do anything with recursive array items, as the merge logic wouldn't run into an infinite loop so it doesn't really matter.
```suggestion
```

Reply:
As for now, without those check the function throws `RangeError: Maximum call stack size exceeded` for tests "should not merge left/right cyclic array"